internal/handshake/crypto_setup_tls.go

```diff
@@ -1,168 +1,474 @@
+	initialAEAD, err := crypto.NewNullAEAD(perspective, connID, protocol.VersionTLS)
 	if err != nil {
-		return nil, err
-	}
-	conn := newCryptoStreamConn(cryptoStream)
-	tls := mint.Client(conn, config)
-	return &cryptoSetupTLS{
-		tls:               tls,
-		conn:              conn,
-		perspective:       protocol.PerspectiveClient,
-		nullAEAD:          nullAEAD,
-		keyDerivation:     crypto.DeriveAESKeys,
-		handshakeEvent:    handshakeEvent,
-		handshakeComplete: handshakeComplete,
-	}, nil
-}
-
-func (h *cryptoSetupTLS) HandleCryptoStream() error {
-	for {
-		if alert := h.tls.Handshake(); alert != mint.AlertNoAlert {
-			return fmt.Errorf("TLS handshake error: %s (Alert %d)", alert.String(), alert)
+		return nil, nil, err
+	}
+	cs := &cryptoSetupTLS{
+		initialStream:          initialStream,
+		initialAEAD:            initialAEAD,
+		handshakeStream:        handshakeStream,
+		readEncLevel:           protocol.EncryptionInitial,
+		writeEncLevel:          protocol.EncryptionInitial,
+		handleParamsCallback:   handleParams,
+		handshakeEvent:         handshakeEvent,
+		handshakeComplete:      handshakeComplete,
+		logger:                 logger,
+		perspective:            perspective,
+		clientHelloWrittenChan: make(chan struct{}),
+		messageChan:            make(chan []byte, 100),
+		receivedReadKey:        make(chan struct{}),
+		receivedWriteKey:       make(chan struct{}),
+	}
+	var extHandler tlsExtensionHandler
+	switch perspective {
+	case protocol.PerspectiveClient:
+		extHandler, cs.receivedTransportParams = newExtensionHandlerClient(
+			params,
+			versionInfo.initialVersion,
+			versionInfo.supportedVersions,
+			versionInfo.currentVersion,
+			logger,
+		)
+	case protocol.PerspectiveServer:
+		extHandler, cs.receivedTransportParams = newExtensionHandlerServer(
+			params,
+			versionInfo.supportedVersions,
+			versionInfo.currentVersion,
+			logger,
+		)
+	}
+	qtlsConf := tlsConfigToQtlsConfig(tlsConf)
+	qtlsConf.AlternativeRecordLayer = cs
+	qtlsConf.GetExtensions = extHandler.GetExtensions
+	qtlsConf.ReceivedExtensions = extHandler.ReceivedExtensions
+	cs.tlsConf = qtlsConf
+	return cs, cs.clientHelloWrittenChan, nil
+}
+
+func (h *cryptoSetupTLS) RunHandshake() error {
+	var conn *qtls.Conn
+	switch h.perspective {
+	case protocol.PerspectiveClient:
+		conn = qtls.Client(nil, h.tlsConf)
+	case protocol.PerspectiveServer:
+		conn = qtls.Server(nil, h.tlsConf)
+	}
+	if err := conn.Handshake(); err != nil {
+		close(h.receivedReadKey)
+		close(h.receivedWriteKey)
+		return err
+	}
+	close(h.handshakeComplete)
+	return nil
+}
+
+func (h *cryptoSetupTLS) HandleData(data []byte, encLevel protocol.EncryptionLevel) error {
+	var buf *bytes.Buffer
+	switch encLevel {
+	case protocol.EncryptionInitial:
+		buf = &h.initialReadBuf
+	case protocol.EncryptionHandshake:
+		buf = &h.handshakeReadBuf
+	default:
+		return fmt.Errorf("received handshake data with unexpected encryption level: %s", encLevel)
+	}
+	buf.Write(data)
```

Comment:
WDYT about factoring wire stuff out into a separate helper / file.

Reply:
#1552 moves this to the quic package.